### PR TITLE
UI tools for VR/AR

### PR DESCRIPTION
**Pull request by moreApi**
- introduces Vr grabbing & moving behaviour
- Extends vr controler example by vr grabbing & moving
- vr volume cropping example using grabbing
- introduce VR zoom behaior
- Extends vr controler example by vr zoom
- introduce VR selection behaior
- Extends vr controler example by vr selection

### CODE REVIEW

**Comment by moreApi** [comment]
Found some todos while presenting it to luke:

- add selection utility function only with lambda no storage 
- investigate ob trigger jetzt funktioniert
- rotation of selection wheel

**Comment by moreApi** [comment]
The orientation of the selection wheel is still a bit strange. But i would suggest merging it nonetheless and improve the orientation later.

File: src/main/kotlin/graphics/scenery/attribute/spatial/DefaultSpatial.kt [hunk 1]
```diff
@@ -1,18 +1,18 @@
 package graphics.scenery.attribute.spatial
 
+import com.jogamp.opengl.math.Quaternion
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/attribute/spatial/DefaultSpatial.kt:3]: This import shouldn't be there.

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    protected val onGrab: (() -> Unit)? = null
+) : DragBehaviour {
+
+    val controllerSpatial: Spatial
+
+    init {
+        controllerSpatial = controllerHitbox.spatialOrNull()
+            ?: throw IllegalArgumentException("controller hitbox needs a spatial attribute")
+    }
+
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:38]: Should all of these vars be public indeed?

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+
+open class Pressable(
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt:83]: Docs missing, should maybe be moved to its own file?

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRScale.kt [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+class VRScale(
+    private val name: String,
+    private val controller: Spatial,
+    private val offhand: VRScaleOffhand,
+    private val setScale: (Float) -> Unit
+) :
+    DragBehaviour {
+
+    protected val logger by LazyLogger(System.getProperty("scenery.LogLevel", "info"))
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRScale.kt:21]: The `System.getProperty()` is not necessary, LazyLogger does that on its own.

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+ * @param cutoff  after this distance between controller and targets no action will be selected if the button is released
+ */
+class VRSelectionWheel(
+    val controller: Spatial,
+    val scene: Scene,
+    val hmd: TrackerInput,
+    var actions: List<Action>,
+    val cutoff: Float = 0.1f,
+) : DragBehaviour {
+    protected val logger by LazyLogger(System.getProperty("scenery.LogLevel", "info"))
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt:28]: Same as above.

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+import graphics.scenery.utils.LazyLogger
+import graphics.scenery.utils.Wiggler
+import org.scijava.ui.behaviour.DragBehaviour
+
+/**
+ * A selection wheel to let the user choose between different actions.
+ *
+ * The list of selectable actions can be changed dynamically.
+ *
+ * @param actions List of named lambdas which can be selected by the user
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt:18]: For Kotlin Kdoc, the `@param` notation should not be used - it's oldschool javadoc. Use `[nameOfTheParameter]` instead 👍

File: src/main/kotlin/graphics/scenery/primitives/LineBetweenNodes.kt [hunk 7]
```diff
@@ -0,0 +1,24 @@
+package graphics.scenery.primitives
+
+import graphics.scenery.attribute.spatial.Spatial
+import org.joml.Vector3f
+
+/**
+ * A line which always goes from one node to another.
+ */
+class LineBetweenNodes(var from: Spatial, var to: Spatial, transparent: Boolean = false, simple: Boolean = false) :
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/primitives/LineBetweenNodes.kt:9]: Not entirely sure this is necessary. You could make a unit line, and then use `Node.orientBetweenPoints`.

File: src/test/kotlin/graphics/scenery/tests/examples/advanced/VRControllerExample.kt [hunk 8]
```diff
@@ -1,35 +1,54 @@
...
+ * Contents:
+ * - Boxes to select and scale and to grab and move
+ * - A grabable spray can to the left
+ * - A touchable party sphere to the right
+ *
+ * Available Controls:
+ * Side buttons alone:  Grab Object
+ * Both side buttons together: Move to scale, after selection
+ * Right Trigger:       Select to Scale
+ * Left Trigger:        Select Party first, then Scale
```
Comment (skalarproduktraum) [inline src/test/kotlin/graphics/scenery/tests/examples/advanced/VRControllerExample.kt:32]: ... Party? 🥳

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
Looks great, only a bit of minor stuff left 👍

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt (hunk 2 above)
Comment (moreApi) [reply src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:38]: they should be protected. There is already a child class who uses this in my microscenery.

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt (hunk 6 above)
Comment (moreApi) [reply src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt:18]: https://kotlinlang.org/docs/kotlin-doc.html#param-name they are using @param here. I don't understand.

File: src/main/kotlin/graphics/scenery/primitives/LineBetweenNodes.kt (hunk 7 above)
Comment (moreApi) [reply src/main/kotlin/graphics/scenery/primitives/LineBetweenNodes.kt:9]: Switched to `Node.orientBetweenPoints`. 
The From and To spatials need to be saved somewhere for the update to work. IMO it's the moste clean solution to save them in an extra class.

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt (hunk 6 above)
Comment (skalarproduktraum) [reply src/main/kotlin/graphics/scenery/controls/behaviours/VRSelectionWheel.kt:18]: Okay, disregard. I think I misunderstood the convention. Keep the `@param` version 👍

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+                    grabable.onDrag?.invoke()
+                }
+            }
+        }
+
+        lastPos = controllerSpatial.worldPosition()
+        lastRotation = controllerSpatial.worldRotation()
+    }
+
+    override fun end(x: Int, y: Int) {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:94]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+        controllerSpatial = controllerHitbox.spatialOrNull()
+            ?: throw IllegalArgumentException("controller hitbox needs a spatial attribute")
+    }
+
+    var selected = emptyList<Node>()
+
+    var lastPos = Vector3f()
+    var lastRotation = Quaternionf()
+
+    override fun init(x: Int, y: Int) {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:44]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+        }
+        if (selected.isNotEmpty()) {
+            onGrab?.let { it() }
+        }
+        selected.forEach {it.getAttributeOrNull(Grabable::class.java)?.onGrab?.invoke()}
+        lastPos = controllerSpatial.worldPosition()
+        lastRotation = controllerSpatial.worldRotation()
+    }
+
+    override fun drag(x: Int, y: Int) {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:57]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+        lastPos = controllerSpatial.worldPosition()
+        lastRotation = controllerSpatial.worldRotation()
+    }
+
+    override fun end(x: Int, y: Int) {
+        selected.forEach {it.getAttributeOrNull(Grabable::class.java)?.onRelease?.invoke()}
+        selected = emptyList()
+    }
+
+    companion object {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRGrab.kt:99]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+
+import graphics.scenery.Node
+import graphics.scenery.Scene
+import graphics.scenery.attribute.spatial.Spatial
+import graphics.scenery.controls.OpenVRHMD
+import graphics.scenery.controls.TrackedDeviceType
+import graphics.scenery.controls.TrackerRole
+import org.scijava.ui.behaviour.DragBehaviour
+
+open class VRPress (
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt:11]: Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+
+    override fun drag(x: Int, y: Int) {
+        selected.forEach {it.getAttributeOrNull(Pressable::class.java)?.onHold?.invoke()}
+    }
+
+    override fun end(x: Int, y: Int) {
+        selected.forEach {it.getAttributeOrNull(Pressable::class.java)?.onRelease?.invoke()}
+        selected = emptyList()
+    }
+    companion object {
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRPress.kt:47]: Empty line before and Kdoc missing

File: src/main/kotlin/graphics/scenery/controls/behaviours/VRScale.kt [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+import graphics.scenery.attribute.spatial.Spatial
+import graphics.scenery.controls.OpenVRHMD
+import graphics.scenery.controls.TrackedDeviceType
+import graphics.scenery.controls.TrackerRole
+import graphics.scenery.utils.LazyLogger
+import org.scijava.ui.behaviour.DragBehaviour
+import kotlin.concurrent.thread
+
+
+class VRScale(
```
Comment (skalarproduktraum) [inline src/main/kotlin/graphics/scenery/controls/behaviours/VRScale.kt:13]: Kdoc missing

**Review by skalarproduktraum — CHANGES_REQUESTED** [review]
Looks great, apart from Kdoc missing in a lot of places, could you please add that? Thanks!

**Comment by moreApi** [comment]
This should be now the best commented branch in scenery. :P

**Review by skalarproduktraum — APPROVED** [review]

**Pull request merged**